server/netty-server/src/main/scala/sttp/tapir/server/netty/example/NettyServerHandler.scala

```diff
@@ -0,0 +1,86 @@
+package sttp.tapir.server.netty.example
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.jdk.CollectionConverters._
+
+import io.netty.buffer.{ByteBuf, Unpooled}
+import io.netty.channel.{ChannelFutureListener, ChannelHandlerContext, SimpleChannelInboundHandler}
+import io.netty.handler.codec.http.{
+  DefaultFullHttpResponse,
+  FullHttpRequest,
+  FullHttpResponse,
+  HttpHeaderNames,
+  HttpHeaderValues,
+  HttpRequest,
+  HttpResponse,
+  HttpResponseStatus,
+  HttpUtil,
+  HttpVersion
+}
+import sttp.tapir.model.ServerResponse
+import sttp.tapir.server.netty.NettyServerInterpreter
+import sttp.tapir.server.netty.NettyServerInterpreter.NettyRoutingResult
+
+class NettyServerHandler(val handlers: List[FullHttpRequest => NettyRoutingResult])(implicit val ec: ExecutionContext)
+    extends SimpleChannelInboundHandler[FullHttpRequest] {
+
+  private lazy val routingFailureResp = {
+    val res = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.NOT_FOUND)
+    res.headers().set(HttpHeaderNames.CONTENT_LENGTH, res.content().readableBytes())
+    res
+  }
+
+  private def toHttpResponse(serverResult: Either[Int, ServerResponse[ByteBuf]], req: FullHttpRequest) = {
+    val resp = serverResult match {
+      case Left(httpCode) =>
+        val error = new DefaultFullHttpResponse(req.protocolVersion(), HttpResponseStatus.valueOf(httpCode))
+        error.headers().set(HttpHeaderNames.CONTENT_LENGTH, error.content().readableBytes())
+        error
+
+      case Right(interpreterResponse) =>
+        val res = new DefaultFullHttpResponse(
+          req.protocolVersion(),
+          HttpResponseStatus.valueOf(interpreterResponse.code.code),
+          interpreterResponse.body.getOrElse(Unpooled.EMPTY_BUFFER)
+        )
+
+        interpreterResponse.headers
+          .groupBy(_.name)
+          .foreach { case (k, v) =>
+            res.headers().set(k, v.map(_.value).asJava)
+          }
+
+        res
+    }
+
+    resp.headers().set(HttpHeaderNames.CONTENT_LENGTH, resp.content().readableBytes())
+    resp
+  }
+
+  override def channelRead0(ctx: ChannelHandlerContext, req: FullHttpRequest): Unit = {
+    if (HttpUtil.is100ContinueExpected(req)) {
+      ctx.writeAndFlush(new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.CONTINUE));
+    } else {
+      //todo: AbstractByteBuf.checkAccessible hack
+      val accessibleRequest = req.copy().replace(Unpooled.copiedBuffer(req.content()))
+
+      val responses: List[Future[FullHttpResponse]] = handlers
+        .map(_.apply(accessibleRequest))
```

Comment:
This will process the request concurrently by all handlers. Instead, we should process them one-by-one, sequentially. Maybe that's the source of the problem with reading the body as well?

Reply:
So I think I got it.

Invoking methods on `req.content()` had side-effects and e.g. `req.content().array()` affected buffer for subsequent
handlers, so doing it for first request was 👍  but for a second one the buffer was spent. It didn't matter if I did it sequentially or in parallel.

I had 3 options:

- copying buffer before processing
- copying underlying array when arriving at `NettyRequestBody`
- manipulating `readerIndex` - marking start before reading with `ByteBuf.markReaderIndex()` and rewinding it after reading with `ByteBuf.resetReaderIndex()`. (Seems like a lot of trouble).

I decided to go with option 2.